### DISCUSSION TITLE
Decipher get_texttrace

### DISCUSSION THREAD

**Discussion by sgthuth** [pymupdf/PyMuPDF#2016] [Looking for help] [answered]
Good day everyone,

Problem: 
   Long story short, I have contend with a program that converts PDF files to another format.  During this conversion, text that was previously not seen on the PDF, is now visible within the new format.  Since I do not have the ability to alter how the other program works, I thought Python would be a good means to solution and I've found a love for PyMuPDF.  
Solutioning:
   At first thought, I tried to convert each page of the pdf to an image; then, build a new pdf with those images.  Works great excluding the file size.  The file size is or at least will be a problem especially with hitting a DPI of 300.  I thought about compression and if compressing the images prior to building the new pdf would work.  I started down that road and of course, I started doing some more crazy things like cutting each page's image into individual pieces to a puzzle but looking at the sizes... it appears all the little image sizes added together equals a size greater than just the single image. I didn't continue with that concept nor work with the images compressed, but I began searching for another solution.  (If anyone knows if the compression approach would work OR knows of a better way of doing that, please let me know).  
   Because the text that becomes visible is under other layers, I started searching for something that could distinguish layers on each page.  That's when I came across get_texttrace [HERE](https://github.com/pymupdf/PyMuPDF/discussions/1814).  I wrote up a short script just to see what the output would look like in a dump.txt file.  What I see is similar to a JSON file format with key value pairs.  I know there are settings for this but I wasn't able to find anything specific to put other than comparisons to the other things like words, blocks, dicts, etc.  

Any advice is much appreciated.

**Comment by JorjMcKie** [comment, marked as answer]
You seem to be resourceful in finding new ideas. I recommend reading the documentation about `get_texttrace()` to find the meaning of the dictionary items.
What else do you need?

**Reply by sgthuth** [reply]
Thanks @JorjMcKie.  Was 'Looking For Help' to see if that is the best way forward, if I should go back to the compression, or maybe there's something entirely different I haven't heard of.  I have read the section.  There's several attributes returned from the method and I was looking for a shove in the right direction.  Like relevance? Application? I haven't touched on things like ascenders, descenders, flags, wmode...

**Reply by JorjMcKie** [reply]
If you need to find out whether text is hidden because of the Text Rendering Mode (PDF command `Tr`) or because it actually just is covered by other objects (e.g. images), the thread you already have been looking at [here](https://github.com/pymupdf/PyMuPDF/discussions/1814) contains everything that can be said about it. You not have looked at `page.get_bboxlog()` to see if your unwanted text is underneath some rectangle coming further down in that bbox list.
There are no other options.

There exists a PDF feature called "Optional Content" (OC), which offers a way to show or hide stuff. This is controlled via special PDF objects ... again: more in PyMuPDF's documentation.

Theoretically, you do not see the text in your original, because its relevant OC state is off. You can check if your PDF is using this feature at all using the more advanced PDF viewers (Adobe, FoxIt, PDF XChange, ...)  under "levels" or "layers". Although MuPDF text extraction respects this too ... and that text shouldn't show up then either.
But you said you have no control over that "other" program, so there is no way for me to judge what it may do.

**Reply by sgthuth** [reply]
I'll have many variations and programs that build the initial PDF files.  I will look into the optional content.  May have to find a means to systematically determine of the PDF file uses the OC feature.  Thanks again, @JorjMcKie

**Reply by JorjMcKie** [reply]
> May have to find a means to systematically determine of the PDF file uses the OC feature.

Easy enough: `if doc.get_layers() == (): False` ... and more about this topic in the documentation.

**Reply by sgthuth** [reply]
Thank you @JorjMcKie .  Do you have an opinion on the imaging route?  Seems a bit straight forward but would in turn have to solve for file size.

**Reply by JorjMcKie** [reply]
> Thank you @JorjMcKie . Do you have an opinion on the imaging route? Seems a bit straight forward but would in turn have to solve for file size.

I am not particularly fond of that one: (1) it is clumsy, (2) the output is mostly unusable except for looking at it (e.g. no more text extraction possible), (3) nothing except the visual page appearances of the original will survive, (4) either bad for zooming or an unattractive file size.

To some extent the file size can be kept under control: of course by choosing the resolution (dpi), and then by potentially **_not_** making color images (which saves a lot of space), maybe plus **_not_** using PNG as the image format but (via pillow) JPEG.

**Reply by sgthuth** [reply]
Hmmm... The main reason I thought the imaging would be best is because the converted format is just being printed.  There's some validation steps that occur in that other program but it is minimal.  I'm scraping the document prior to creating new PDFs based on images.  Any information needed on the document is being retained by me or placed back on there but all in all... it's just going to be printed.  When I was on the idea, I found [THIS](https://github.com/pymupdf/PyMuPDF/issues/393).  I'm assuming with the age of that discussion, some methods have been updated (getPixmap -> get_pixmap).  I tried `ppm = pix.getImageData('ppm')` but couldn't get it to take.

**Reply by JorjMcKie** [reply]
Ah ok. But if printing is the only purpose:
1. The logic needed to make a page appearing as it should is unbeatably easy with making pixmaps of course
2. Why does PDF size matter so much then - the output appears to be just a through-away product!
3. You just must be careful with choosing the resolution
4. You seem to have a thorough dislike to read documentation (😉): there is an extra chapter / section dealing with the major renaming effort two years ago.
5. This `jpeg = pix.pil_tobytes("JPEG", ... more pillow parameters ...)` will create `bytes` in JPEG image format using PIL / Pillow internally. The pillow documentation will also have hints on how to optimize / minimize the resulting binary. Using "ppm" output is about the largest (although fastest) alternative.

**Reply by sgthuth** [reply]
Well as much as the throw away concept should apply, size is a major consideration.  The secondary application that converts PDFs is also concatenating them in a single file.  It could be just a few PDF files that range from 1 to 5 pages each.  It could be upwards 5000 PDF files but on average, it's about 500 PDF files at a time.  I experimented with the image concept with our considering compression just to see.  The resulting culmination of PDFs was about 40 times the original size and the converted format was about 10 times.  I'd be happy if I can just get it under 5 times.  I like to read and learn.  Just a matter of finding the right pieces to read and tie them together to create the best solution.  One day I'll read through all the documentation.  I'd like to build something similar to manipulate AFP files which is the resulting format in this process that I have to contend with.  Needless to say about a month ago, I'd only used Python in lambda functions for AWS and had never thought about the structure of a PDF much-less heard of PyMuPDF.  You guys have done an amazing job... and yes, it has been well documented.  I guess I'll get to reading.

**Reply by JorjMcKie** [reply]
Well the documentation admittedly is a huge book in v1.20.x already: 420 Letter pages.
But the packages on the other hand is a large one too.

Just looking at an arbitrary PDF page with text and one image:

```python
len(page.get_pixmap().tobytes())
202299
len(page.get_pixmap().pil_tobytes("jpeg",optimize=True))
81306
```

**Reply by sgthuth** [reply]
Oh wow.  That is very helpful!  Thank you @JorjMcKie .  I'll come back over here and post what I've done but that's plenty to go in that direction.  I'll read up, too!  Thanks for all the help once again.

**Comment by sgthuth** [comment]
Alright... Either my brain is fried or I'm just having a bad day... 
I've gotten 95% of my code written out but I can't recall or find (again) how to append the current page from my template with no alterations 'as is' to my final pdf.  I think my brain is just not working.  A bit of my code is below.

```python
with fitz.open(path_and_filename_of_original_pdf) as original_pdf:
   final_pdf = fitz.open()
   for p, page in enumerate(original_pdf, 0):
      rdt = [] # list of redactions
      adt = [] # list of additions
      blocks = page.get_text('blocks')
      blocks.sort(key=lambda block: block[1])
      for b in blocks:
         if b[4].contains(x):
            rdt.append(b[4].strip())
            if c:
               cpl = True # Cripple PDF
         if b[4].contains(y):
            adt.append(fitz.Point(b[0], b[1]), b[4].strip())
      for r in rdt:
         draft = page.search_for(r)
         for fnd in draft:
            page.add_redact_annot(fnd)
            page.apply_redactions()
            page.apply_redactions(images=fitz.PDF_REDACT_IMAGE_NONE)
      if cpl:
         if len(blocks) > 0:
            pix = page.get_pixmap(matrix=fitz.Matrix(4.16, 4.16))
            jpeg = pix.pil_tobytes("JPEG", optimize=True, mode="L")
            imgPDF = fitz.open("pdf", jpeg)
            imgPage = final_pdf.new_page(width=rect.width, height=rect.height)
            imgPage.show_pdf_page(rect, imgPDF, 0)
         else:
            final_pdf.new_page(width=rect.width, height=rect.height)
      else:
         # PASS PAGE AS-IS TO FINAL PDF
      for a in adt:
         page.insert_text(a[0], a[1], color=(), fontsize=7, fontname="Times-Roman")
   final_pdf.save(path_and_filename_of_final_pdf, garbage=4, deflate=True, clean=True)
```
A little hint to pass the current state of a page to the final pdf?

**Reply by sgthuth** [reply]
If iterating over the pages in the template and making changes, will the following capture the page's current state?

```python
final_pdf.insert_pdf(template_pdf, from_page=p, to_page=p)
```

**Reply by sgthuth** [reply]
Was just trying to experiment and see if the insertion would work but I can't get passed...
```Python
imgPDF = fitz.open("pdf", jpeg)
```
Exception Thrown: 
cannot open broken document

**Reply by sgthuth** [reply]
Got it passed imgPDF... had to change a bit...

```Python
# Cripple Page
pix = page.get_pixmap(matrix=fitz.Matrix(4.16, 4.16))
jpeg = pix.pil_tobytes(format='jpeg', optimize=True, mode='L')
newPage = final_pdf.new_page(width=rect.width, height=rect.height)
newPage.insert_image(rect, stream=jpeg, overlay=False)
```

is mode "L" not grayscale for pillow?  Somehow my resulting PDF has been updated to images but it's still in color.  Lol.

Moving on to processing a PDF that doesn't need crippling to determine if insertion will work or not to pass page 'as-is'

**Reply by JorjMcKie** [reply]
you must use colorspace gray already when making the pixmap: `pix=page.get_pixmap(colorspace=fitz.csGRAY, ...)`.

**Reply by sgthuth** [reply]
Thanks @JorjMcKie!

I got it late last night.  All things cripple are working as anticipated.  Only issue now is when crippling is not required... I'm going to try the insert_pdf method.  Hoping there's a simple way to appending the current page and it's current state within the original pdf's page iteration to the end of the final pdf.

```Python
final_pdf.insert_pdf(original_pdf, from_page=p, to_page=p)
```
